### ISSUE TITLE
i10D personality module not identified correctly

### ISSUE THREAD

**Issue by george-basco** [tagyoureit/nodejs-poolController#207]
**Describe the bug**
I have an Intellicenter with the i10D (523029Z) module.  It is incorrectly being identified as a 522039 cover module.  The system is being identified as an i5P.

**Expected behavior**
initExpansionModules is being called with:
ocp0A=0x87, ocp0B=0, ocp1A=0, ocp2A=0, ocp3A=0

The slot1 valve expansion board gets identified correctly.
slot0 (value of 7) should be identified as i10D but is not.

I changed the code in IntellicenterBoard.ts so that a value of 7 identifies the i10D expansion board.
And also changed the line:
  if (pb.type === 0 || pb.type > 5)
that identifies the system as i5P to:
  if (pb.type === 0 || pb.type > 7)

I then get correct identification in poolConfig.json, but the rest of the config isn't being filled out, e.g. circuits.  I'm looking at this code for the first time, so don't understand at all what's going on.

Let me know what information would be helpful.  Thanks..

**Pool Equipment**
 - Controller: Intellicenter
 - Controller Model: IntelliCenter i10D
 - Controller Firmware Version: 1.045
 - Pump(s) manufacturer and model: IntelliFlow 022056
 - Valves: Intellivalve

**Comment by george-basco** [comment]
Here's what poolConfig.json looks like before my couple of changes:

{
  "lastUpdated": "9/2/2020, 18:30:01",
  "pool": {
    "location": {
      "timeZone": 139
    },
    "options": {
      "units": 0,
      "adjustDST": true
    }
  },
  "equipment": {
    "maxBodies": 0,
    "maxCircuits": 0,
    "maxSchedules": 100,
    "maxPumps": 16,
    "maxValves": 6,
    "maxCircuitGroups": 16,
    "maxLightGroups": 40,
    "maxChlorinators": 0,
    "maxChemControllers": 0,
    "maxFeatures": 32,
    "model": "IntelliCenter i5P",
    "tempSensors": [
      {
        "id": "air",
        "isActive": true,
        "calibration": 0,
        "name": "Air"
      },
      {
        "id": "water1",
        "isActive": true,
        "calibration": 0,
        "name": "Body 1"
      },
      {
        "id": "solar1",
        "isActive": false,
        "calibration": 0,
        "name": "Solar 1"
      },
      {
        "id": "water2",
        "isActive": false,
        "calibration": 0,
        "name": "Body 2"
      },
      {
        "id": "solar2",
        "isActive": false,
        "calibration": 0,
        "name": "Solar 2"
      },
      {
        "id": "water3",
        "isActive": false,
        "calibration": 0,
        "name": "Body 3"
      },
      {
        "id": "solar3",
        "isActive": false,
        "calibration": 0,
        "name": "Solar 3"
      },
      {
        "id": "water4",
        "isActive": false,
        "calibration": 0,
        "name": "Body 4"
      },
      {
        "id": "solar4",
        "isActive": false,
        "calibration": 0,
        "name": "Solar 4"
      }
    ],
    "modules": [
      {
        "id": 0,
        "name": "A/D Module",
        "desc": "A/D Cover Module",
        "type": 7,
        "part": "522039",
        "covers": 2
      },
      {
        "id": 1,
        "name": "Valve Exp",
        "desc": "Valve Expansion Module",
        "type": 8,
        "part": "522440",
        "valves": 6
      }
    ],
    "expansions": [
      {
        "id": 1,
        "modules": []
      },
      {
        "id": 2,
        "modules": []
      },
      {
        "id": 3,
        "modules": []
      }
    ],
    "shared": false,
    "dual": false,
    "equipmentIds": {
      "circuits": {
        "start": 2,
        "end": 4
      },
      "features": {
        "start": 129,
        "end": 160
      },
      "circuitGroups": {
        "start": 193,
        "end": 208
      },
      "virtualCircuits": {
        "start": 237,
        "end": 292
      },
      "invalidIds": []
    },
    "softwareVersion": "1.045"
  },
  "configVersion": {
    "equipment": 0
  },
  "bodies": [],
  "schedules": [],
  "circuits": [],
  "features": [],
  "pumps": [],
  "chlorinators": [],
  "valves": [],
  "heaters": [],
  "covers": [],
  "circuitGroups": [],
  "lightGroups": [],
  "remotes": [],
  "security": {},
  "customNames": [],
  "eggTimers": [],
  "chemControllers": [],
  "appVersion": "6.0.1",
  "general": {},
  "controllerType": "intellicenter"
}

**Comment by george-basco** [comment]
[replay (6).zip](https://github.com/tagyoureit/nodejs-poolController/files/5165161/replay.6.zip)

**Comment by tagyoureit** [comment]
@rstrouse can you take a look?

**Comment by rstrouse** [comment]
Ahah, we finally see the signatures for i10D.  I have modified the code to see your signature.  This has been published to next.  The dual equipment configuration is new so let me know if there are any other issues.  The place to look is with the valve and heater attachments.  There was a change with 1.047 beta to the heat sources and heat modes so let me know if these are not correct in 1.045.  The official release is still 1.040 and I don't know what Pentair did with these in 1.045.  It's starting to look like there are more people running unofficial releases than official ones.

**Comment by george-basco** [comment]
I had already tried the change you made.  Also need to change this line:

https://github.com/tagyoureit/nodejs-poolController/blob/f11816b9c666e6417da34014d301339cfa5d85f5/controller/boards/IntelliCenterBoard.ts#L296

Those changes did identify the i10D correctly, but didn't do anything to identify pumps, valves, circuits, etc.   Currently none of those things are getting filled in.  If you look at the zip file above you'll see most things are empty.  To start with there are two pumps and 6 valves attached that aren't showing up anywhere.  Not sure where to start..

**Comment by rstrouse** [comment]
Oh yes there is a some goofiness with the i5p as this has no board installed in slot 0 but it reports garbage in the MSB of the lower nibble.  I think that only the first 3 bits are defined for the slot.  The other equipment should be detected as soon as the correct board signature is detected.  I assume that you get a total of 10 valves for i10D + valve expansion.  Is there a valve dedicated for each body or how does that work since there is no intake/return thingy going on?  I posted up changes to drop that bit from the slot0 i5P detection.

Once the software starts you will see the njspc start to ask for the configuration items.   It only asks for the items that it needs and since the definition didn't include all the equipment counts from the signature it didn't ask IntelliCenter for the equipment defintitions.  You should see a line that looks says something like Queued 70+ Configuration items for your configuration at startup.  After it acquires all the equipment it will only ask for the things that have changed since the last time it asked.  That is what the configVersion section is all about in the poolConfig.json.
![image](https://user-images.githubusercontent.com/47839015/92183191-8102a180-ee02-11ea-93fc-f19500db8fb6.png)

**Comment by rstrouse** [comment]
Once it detects the i10D correctly do another capture as your packetLog shows the OCP not responding to action 228.  The OCP is supposed to respond with the versions of currently installed equipment.  This is perhaps because the requesting address was not properly being set because of the lack of personality board being found.

**Comment by george-basco** [comment]
i10D now correctly gets identified, but still no response for action 228.  Here are the logs:

[replay (8).zip](https://github.com/tagyoureit/nodejs-poolController/files/5172688/replay.8.zip)

**Comment by rstrouse** [comment]
It sounds like you are pretty technically adept.  I have some questions and I am confident that we can figure this out.  It is odd that it is not responding to 228.  However, this is not to say that they didn't change something between 1.040 and 1.045 that might make a difference here.  If you have installed dashPanel I would like you to load up the messageManager that is included in this package.  It can be opened by typing ```http://<ipaddress of dashPanel server>/messageManager.html```  Bear in mind these are the tools that we use to inspect the RS485 bus and haven't spent too much time documenting them.  In the end I want to see what is communicated on this bus from other slave controllers.

What this software does is monitor the RS485 bus traffic.  The screenshot is a picture of your log you sent me.  From this log I see that the bus is healthy but the OCP is just not responding to the 228 action -- "Request for Equipment Versions."  This log also tells me that njspc is correctly now awaiting responses from the panel on what to do next.  I can see the OCP taking control of 3 pumps on the bus and sending information related to the temperature sensors, circuit states, and other operations.

The typical response to 228 is action 164 which lets njspc know what equipment is installed on the controller.  Do you have an indoor controller or a wireless controller connected to the RS485 bus.  If you do there are tells from these controllers.  It could be as simple as finding the correct request source for the message.  We can find this by observing what the other controllers on the bus are asking for and how they are asking for it.   Once you get the messageManager up and running go to the next post for further instructions.
![image](https://user-images.githubusercontent.com/47839015/92269182-8ca0a780-ee98-11ea-8ff1-f7a83f9b84a1.png)

**Comment by rstrouse** [comment]
Your mission, should you choose to take it is to infiltrate the RS485 communications, retrieve the secret discussion taking place under our nose and report back here.  Unfortunately, this is where the drama ends and this message won't self destruct.

Anyhow, once installed, click the upper right corner where the menu bars are located.  This will open a panel that contains the connection and the logging options for the messageManager.   If the screen is showing "Not Connected' then we need to set the ip address for njspc into the connections tab.   Once we are connected, click on the logging tab.  From there I would like to filter out some of the messages that are coming in so we don't get so much noise.  We are only interested in broadcast messages so click the Packets button, File button, and the Broadcast button.  Everything else is just noise at this point.  Your filter should look like the screenshot.
![image](https://user-images.githubusercontent.com/47839015/92271971-911b8f00-ee9d-11ea-97cf-dfaae8299d51.png)


Once you have this set up you can start observing the messages on the RS485 bus.   It would be spectacular to see a 228 message that wasn't sent by njspc, but if we don't that's ok.  Click the last button on the top of the messages list (make it pink) to begin receiving messages on the bus.  You should start seeing action 2 and 204 being sent at regular intervals.  Then wake up an indoor control panel (if you have one).  If you don't then use the web or mobile app and change a configuration setting.  Simply changing the name of a feature will suffice.  Once you have done that you can go back to the Logging tab and uncheck Log to -- File button.  This will stop writing messages to the file.  Please remember to do this as that file can grow out of control if you let it.  You can also unclick the message feed on the messages list.

A log file will be written to the njspc logs directory.  Please upload that log here.  It will be timestamped with the date and time that the first message was logged.

**Comment by george-basco** [comment]
Might be doing something wrong, but not getting anything very interesting by changing a circuit name.  Just see a 168 action. 
 Here are some various scenarios I captured.

rename a circuit twice on the IOS app:
[ios-app-rename.log](https://github.com/tagyoureit/nodejs-poolController/files/5177178/ios-app-rename.log)

rename a circuit twice on intellicenter.com web interface:
[web-rename.log](https://github.com/tagyoureit/nodejs-poolController/files/5177182/web-rename.log)

wakeup ICP and rename circuit twice:
[icp-rename.log](https://github.com/tagyoureit/nodejs-poolController/files/5177183/icp-rename.log)

Power cycle the ICP:
[icp-power-cycle.log](https://github.com/tagyoureit/nodejs-poolController/files/5177184/icp-power-cycle.log)

Reboot the OCP:
[ocp-reboot.log](https://github.com/tagyoureit/nodejs-poolController/files/5177185/ocp-reboot.log)

Haven't really looked at them in any detail, but I think only the last two have any interesting data.

Let me know if there are any other scenarios that would be helpful to capture.  Thanks for looking at this.  This Pentair stuff is completely new to me, so clueless on what any of this means.

**Comment by rstrouse** [comment]
Well, look at that.  It is communicating on address 32.   Hmmm, add the pluginAddress line to your config.json file for njspc under the controller section.  Then restart njspc.  Let's see if the i10D recognizes the address.

```
"controller": {
   "pluginAddress": "32",
   "comms": { ...
```

Also, with the message feed turned on and the following filter set up by clicking on the filter icon in the Broadcast button let's send the i10D a request.
![image](https://user-images.githubusercontent.com/47839015/92291184-41a08780-eecc-11ea-973e-7a7fe43d6cc9.png)

# Step 1
On the "Send Message Queue" section of the messageManager  click on add message and enter the information as you see it on this screen.  Then press the Save Message Button.
![image](https://user-images.githubusercontent.com/47839015/92291427-719c5a80-eecd-11ea-8a9b-fadd31d1b5c6.png)

# Step 2
After you press the save message you should see the new message added to the queue on the Send Message Queue section.  Turn on the message feed by making the top right icon.  Then press the Send Queue button.  Since we have all the other messages filtered out we should only see the ones we are interested in.
![image](https://user-images.githubusercontent.com/47839015/92291528-043cf980-eece-11ea-8f8e-69f5f713b080.png)

# Step 3
The expectation is to see a 228 get sent followed immediately by a 164 message.  If the 164 comes back it will trigger njspc to start asking for information from the ocp if it doesn't have it.
![image](https://user-images.githubusercontent.com/47839015/92291605-5251fd00-eece-11ea-9fc9-6d9a0b738c7b.png)

**Comment by george-basco** [comment]
Pretty sure I did this correctly.  There is no response from the OCP to the 128 request.  Nothing but the 2/204 chatter.

![Capture](https://user-images.githubusercontent.com/70659449/92301538-8bfc2580-ef19-11ea-85b5-d1e902b7f7ae.PNG)

**Comment by rstrouse** [comment]
Yes you did this correctly.  Hmmm, change the controller byte from 63 to 1 and try it again.  You can do this by clicking the little edit icon next to the message in the queue.  Then send it again.  If this is true 1.045, changed the controller byte and the source byte.  It will be interesting if they did it per platform, either way it will be a short task to fixing it once we know what that is.

**Comment by george-basco** [comment]
Changing controller from 63 to 1 had no effect.   I'll go back to 1.040 and see what happens with that..

**Comment by rstrouse** [comment]
Will it even let you go back to 1.040 from the 1.045 beta?  Here is something interesting.  Below was from your reboot-icp.log.  Is this rebooting an Indoor Control Panel?  It actually succeeded on the reboot of the control panel in sending back a 164.
![image](https://user-images.githubusercontent.com/47839015/92312392-ee364400-ef74-11ea-895f-79dda468fcf7.png)

Try sending this message to see if we get anything back.  Note the destination is 16 not 15.  Hit send queue a few times just in case we have a collision.
![image](https://user-images.githubusercontent.com/47839015/92312439-63097e00-ef75-11ea-8f86-191131ffb29f.png)

**Comment by george-basco** [comment]
Looks like you can install whatever version of firmware you'd like.  I went back to 1.040.  However, nothing has changed.   Seems like everything from njspc is ignored.

The icp-power-cycle.log file is the case of power cycling the indoor control panel.  If I copy the packet you highlight above and replay it nothing happens.  Btw, going back to 1.040 firmware, I still see the same 228 request go out when the ICP power cycles.  Haven't noticed any differences in the packets between 1.045 and 1.040.

Then I tried replaying 168 actions for a circuit rename from the IOS app.  Again, it seems to get ignored and the name is not changed.

I was stating to think there is something wrong with the RS-485 output from the rpi.  But after looping it back, seems like it's working OK.  And certainly the reading of the bus looks fine.  I haven't captured njspc initiated packets, but simple echo test was good.

Again, seems like everything from njspc is simply ignored, even those packets that are copied and replayed from things like the ICP.  Maybe you have other suggestions on tests that would show the njspc packets are making it to the wire.

The dashPanel interface did manage to populate for the first time when I downgraded the firmware on the OCP and it booted.  And the data looked mostly complete.  However, if I tried to do anything on dashPanel, all actions failed after retrying the command 4 times.

**Comment by george-basco** [comment]
One question I forgot to ask in the previous post is whether you've seen any evidence that something has responded to a packet from njspc in the various logs of mine that you've looked at?

**Comment by rstrouse** [comment]
I have seen no evidence of the OCP responding to any request from njspc.  It's odd in that you even sent it the exact stream of bytes that were sent from the ICP and nada.  Perhaps there is some issue with the RS485 dongle.  It is odd that it would read but not write.  Interestingly, njspc is not particular about which controller asked for the configuration so it probably scraped the data from the ICP asking for it which is how it picked up the configuration.  This has to be a connection problem either with the dongle or the wiring.

Give me a bit of background on what platform you are on and which dongle you are using.  I have had a dongle go stupid after being connected to socat but I suspect this is due to the fact that node serial port doesn't reset all the registers.  I had to force a settings reset it through a PC.  In that case it was acting like it was full duplex (don't even know how that is on RS485 unless there are 2 pairs) but the other way around where it would write but not read.  The weird part was that the dongle only supported half duplex anyway.  In this instance you should only have 2 wires connected up to the dongle.

**Comment by george-basco** [comment]
I have this dongle based on the CP2102 chip:
https://smile.amazon.com/gp/product/B078W5L8W1/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1

Have just the TxD and RxD pins connected to the OCP main board along with the pumps.  The dongle is connected to a rpi 4b running Raspbian GNU/Linux 10 (buster).  Linux raspberrypi 5.4.51-v7l+ #1333 SMP Mon Aug 10 16:51:40 BST 2020 armv7l GNU/Linux.

Was wondering if there was something not quite right with the tty settings, but looks like the interesting tty options are configured in the serial configuration code.  Suppose I should have something else snoop the RS485 to look for the njspc output.  Probably should just order one of the dongles that is more of a known quantity, I ordered mine before doing much research.

**Comment by rstrouse** [comment]
I couldn't see anything in there about supporting RS485 half duplex.  It states that it supports TTL but no discussion about putting it into RS485 mode.  RS485 is different than TTL but I have seen some TTL adapters that can be coerced int o looking at differential voltage rather than TTL which is just on/off.  Our output on TTL will end up in a bucket of bits on some unknown island if we encode it on the wire as non-differential voltage which appears to be happening.

Here is the adapter I have.  It's cheap and it works.
https://www.amazon.com/JBtek-Converter-Adapter-ch340T-Supported/dp/B00NKAJGZM/ref=sr_1_8?dchild=1&keywords=rs485+usb&qid=1599353709&s=electronics&sr=1-8

**Comment by george-basco** [comment]
Oh, I'm an idiot.  Really sorry for wasting your time on this.  Just ordered a proper adapter...

**Comment by rstrouse** [comment]
No worries.  We did manage to figure out a mystery that we had regarding the i10D signature.  I'm excited to see if we made all our assumptions about that board correctly.

**Comment by george-basco** [comment]
Things work much better when you know how to speak RS-485 :-)  Thanks for helping me sort that out.

At first glance most things look correct and I'm able to control things as expected.  The one issue I do see is with the valves.  I forgot to answer your earlier question about valves.  The i10D only has two valves A & B.   With the valve expansion board (6 valves) I have a total of 8.  It also appears that the non-shared personalities all have the two valves.  This include i5P, i8P, i10P and i10D.  You can see this on page 61:  https://www.pentair.com/content/dam/extranet/pentair-pool/residential/manuals/automation/intellicenter/intellicenter-automation-system-installation-guide.pdf

Looks like the code in controller/comms/messages/config/ValveMessage.ts will need to be reworked given the lack of inake/return valves.  The current code results in the following dashPanel display for valves:

![image](https://user-images.githubusercontent.com/70659449/92564520-27e5a400-f22e-11ea-872f-d2c0e929501a.png)

vs. what it looks like from intellicenter.com:

![image](https://user-images.githubusercontent.com/70659449/92560288-01703a80-f227-11ea-90dd-91a403bf6e2f.png)

Here are the logs: [replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/5193214/replay.9.zip)

**Comment by rstrouse** [comment]
Glad to see you deciphered the babel.  I didn't realize you could read RS485 with a TTL but that does make sense.

Valves are odd and now that you are here we now know that it does not include either of the intake/return valves in the valve count.  I'll make the mods to the code.  This also means that there is no fancy code for body management unless the water is shared between the two.  Unfortunately, this also leaves the second Intake/Return slot unaccounted for,  We had assumed that Pentair gave you one or the other but I guess the relay to activate it would have to be on the personality board anyway.

I'll fix this this week and post back here when the code is up.  Let me know if there are any other anomalies.

**Comment by rstrouse** [comment]
Alright this week is here.  I posted the fixes for the i10D.  Consequently, your configuration solved some mysteries with the valves but created others.  There are 2 valve positions in the midst of the configuration that are unaccounted for in position 5 and 6 no matter what might be installed.  

Funny, we do have people running single body systems but they haven't reported two extra valves that don't do anything.  Clearly there are no additional valve connections on the P models like there are on the PS.   The same can be said with the X models.  

The claim is that there are 26 supportable valves.  However when I add them up I come up with 28 if these two ghost positions are included.  If you remove them from the count, then the numbers come out correct for 3 expansion panels and 4 valve expansions.  Given the current number of installable boards you can count to 26 physical connections and the position 5 and 6 can never be used.

Anyhow, once you install the update, go to the hamburger menu.  On the system tab, click the "Reload Config" button an this will interrogate the panel again.  The percentage bounces around as njspc detects new equipment to ask for, but once it is done, your valve installations should be correct.  Let me know if there are any other discrepancies.

I did notice something odd about your timezones.  While IntelliCenter has the correct timezone, I suspect the computer that is running the njspc server does not.  If it is a pi these come set to GMT by default with DST turned on.  The DST setting is correct but you may see some weirdness with the time in njspc as we try to localize it with IntelliCenter.

**Comment by george-basco** [comment]
Last two valves are missing names:

![image](https://user-images.githubusercontent.com/70659449/92668426-a3ddfb80-f2c3-11ea-9e29-f9874564e8e1.png)

Reloaded multiple times, but names always missing.

[replay (10).zip](https://github.com/tagyoureit/nodejs-poolController/files/5198511/replay.10.zip)

Timezones should match now..

**Comment by rstrouse** [comment]
Ahah, you two gaps in the names.  This means that it needed to ask for 2 more names.  I have modified the code to ask for these two as well.

**Comment by george-basco** [comment]
Valves all look great now.  Thanks!

I've noticed another issue, which btw doesn't seem important to my setup, with feature circuit type.  All of my feature circuits are of type generic.  However, njspc shows them all as type spillway.  

![image](https://user-images.githubusercontent.com/70659449/92676097-b9f4b780-f2d5-11ea-8a30-6fa3b235bb87.png)

If I change the type in any of the interfaces, njspc tracks the type correctly.  However, njspc seems to not correctly determine their type at startup/reload config, always showing them as type spillway.

I don't know what spillway type even means, but pointing it out in case it's meaningful to others..

**Comment by george-basco** [comment]
Btw, looking at that last screenshot reminded me that the 24h0m egg timer is misleading.  It's a dashPanel issue, but that should be displayed as "don't stop"

**Comment by rstrouse** [comment]
That's funny.  I'm guessing that the type is not relevant to an i10D as spillway is a feature the rotates the Intake and Return such that water spills out of the spa and into the pool.  I'll have a look at the action 30 [5, ...] message.  

I also agree with the don't stop it just comes from the OCP as 24.  We'll translate that into don't stop.  They use a checkbox on the IntelliCenter app but it really is 24 in the hours byte of teh egg timer.

**Comment by rstrouse** [comment]
If you pull both njspc and dashPanel the don't stop function now works and the feature type also works.  You will need to reload config to see the changes.

What options do you see in the IntelliCenter web for feature circuit types?  I see two Generic and Spillway.  Is your list different?
![image](https://user-images.githubusercontent.com/47839015/92799411-9e86bc80-f368-11ea-9feb-a0a79546f62f.png)

**Comment by george-basco** [comment]
All looks and works great!  I see the same list of circuit types as you.  Thanks..

**Comment by rstrouse** [comment]
Strange.  Spillway would only make sense in a shared body system.

**Comment by george-basco** [comment]
Found a problem with light groups.  Not sure if this is related to the dontstop changes or not.  If I try to do anything to a light group in dashPanel, njspc dies.   E.g. if I edit the name of the group:

[9/10/2020, 15:51:14] info: [15:51:14] 192.168.1.28 GET /config/options/lightGroups?null {}
[9/10/2020, 15:51:26] info: [15:51:26] 192.168.1.28 PUT /config/lightGroup {"id":200,"type":1,"name":"All","dontStop":false,"circuits":[{"circuit":3,"color":0,"swimDelay":0,"position":1},{"circuit":4,"color":0,"swimDelay":0,"position":2}],"eggTimer":720}
[9/10/2020, 15:51:26] info: {"id":107,"valid":true,"dir":"out","proto":"broadcast","pkt":[[],[255,0,255],[165,1,15,33,168,40],[6,0,7,1,1,0,2,3,255,255,255,255,255,255,255,255,255,255,255,255,255,255,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,12,0],[15,184]],"ts":"2020-09-10T15:51:26.686-0700"}

/home/pi/src/nodejs-poolController/nodejs-poolController-next/controller/boards/IntelliCenterBoard.ts:1466
                            sgroup.eggTimer = group.eggTimer = eggTimer;
                                           ^
TypeError: Cannot set property 'eggTimer' of null
    at Outbound.onComplete (/home/pi/src/nodejs-poolController/nodejs-poolController-next/controller/boards/IntelliCenterBoard.ts:1466:44)
    at SendRecieveBuffer.clearResponses (/home/pi/src/nodejs-poolController/nodejs-poolController-next/controller/comms/Comms.ts:343:73)
    at SendRecieveBuffer.processInbound (/home/pi/src/nodejs-poolController/nodejs-poolController-next/controller/comms/Comms.ts:404:37)
    at SendRecieveBuffer.processPackets (/home/pi/src/nodejs-poolController/nodejs-poolController-next/controller/comms/Comms.ts:219:21)
    at Timeout._onTimeout (/home/pi/src/nodejs-poolController/nodejs-poolController-next/controller/comms/Comms.ts:192:109)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)

**Comment by rstrouse** [comment]
Pull on njspc.  This was always lurking but the don't stop code exposed it.

**Comment by rstrouse** [comment]
Hey, am I reading this correctly?  Do you have an A/D Cover module installed?

**Comment by george-basco** [comment]
I have a pool cover and was planning to get the cover module.  I played around with the cover config stuff to see what's possible.  After realizing how limited the latest state of the art Pentair programming model is, I started down the RPi path.  So unlikely to get the module at this point and just use the RPi for what I need.

The light group crash is fixed, but I'm still seeing some strangeness with creating/editing light groups.  I think it mostly comes down to the njspc's cached state being out of sync.  Nothing critical, but I like to document any anomalies.

Here's an example using dashPanel:

1. create a new light group named All
2. add two light circuits

![image](https://user-images.githubusercontent.com/70659449/92868038-2646e800-f3b6-11ea-9bf9-e8e8387cce73.png)

3. save the group

![image](https://user-images.githubusercontent.com/70659449/92868150-424a8980-f3b6-11ea-89c9-6c98842a92df.png)

4. Notice the egg timer shows 0h 12m after the save happens
5. Look at poolConfig.json:
```  "lightGroups": [
    {
      "id": 200,
      "type": 1,
      "circuits": [
        {
          "id": 1,
          "circuit": 3,
          "swimDelay": 0,
          "color": 0
        },
        {
          "id": 2,
          "circuit": 4,
          "swimDelay": 0,
          "color": 0
        }
      ],
      "name": "All"
    }
  ],
```
6. Notice no value for **eggTimer**.  Requires a **Reload Config** to get things in sync.  After the reload, poolConfig.json now has:

```
  "lightGroups": [
    {
      "id": 200,
      "type": 1,
      "circuits": [
        {
          "id": 1,
          "circuit": 3,
          "swimDelay": 0,
          "color": 0
        },
        {
          "id": 2,
          "circuit": 4,
          "swimDelay": 0,
          "color": 0
        }
      ],
      "name": "All",
      "isActive": true,
      "lightingTheme": 0,
      "eggTimer": 720,
      "dontStop": false
    }
  ],
```
And dashPanel now shows the expected values:
![image](https://user-images.githubusercontent.com/70659449/92869914-29db6e80-f3b8-11ea-8d2e-85a3f5e15744.png)

**Comment by rstrouse** [comment]
Ah, that explains the empty slot.  I thought were were missing a byte somewhere.  I'll have a look at the light group settings.  Your definition was actually missing several other properties.

You can load the relay equipment manager to control a relay.  It does directly connect to the websocket for njspc as a trigger for the the relay to control the cover.  Do you know how the cover functionality is supposed to work?  Does it disable the listed circuits so that they won't do anything if the cover is closed or does it open the cover then start the circuit?

**Comment by george-basco** [comment]
I don't really know how the cover module works.   There doesn't appear to be much of any documentation for it.  My guess from the interllicenter.com interface is that you can disable circuits in either open/closed state.  Don't think they would support automating the opening/closing of the cover.

One other comment about lights since we;re on the topic, although this would be a dashPanel thing.  I think the various Pentair interfaces (intellicenter.com, IOS app, OCP) all automatically display any light circuits in a "Lights" UI  and provide an all on/off button.   This happens without creating any light groups.  Given dashPanel has a **LIghts** panel on the home screen, might make sense to do the same there.  Not really needed of course, the lights show up in the **Features** panel and you can of course create a light group for the all on/off functionality.  But just a thought considering the Pentair interfaces all seem to do something like this.  Although as I type this I realize that's probably reason to not do it :-)

**Comment by rstrouse** [comment]
Yeah that does seem odd to me though given the architecture.  I assume that the circuit list is somehow disabled so when it gets a request that request is ignored.  There must be some sort of response from the OCP that says it got the request but rejected it.  The constructs obviously don't exist without the module.  Otherwise the controller would try multiple times on each failure.

Btw the group initialization method has been fixed.  The funny part was that it wasn't happening to me because I have many controllers (including 2 njspcs) on the same bus.  These were adept enough to ask for information they didn't have so that communication initialized the egg timer.  Anyhow, it is fixed and thanks for the detailed description.

On the all on/off I have plans on adding it but just haven't figured out what the widgetry will look like.  I have gone back and forth with adding it to the header but there are other things that appear there on *Touch systems for IntelliBrite.  Most notably, in *Touch these items are promoted to the header because you cannot control the light circuits individually.

There is one thing to note about lights and "Show In Features."  They are odd, while you can create features and circuits that do not appear on the interface, you cannot create lights that do not show on the interface.  With lights, the Show In Features checkbox determines whether it shows up in the features list or on the lights list.   This is because some of the IntelliBrite based control elements are both lighting and feature.  This includes the lighted water feature equipment like MagicStream and the bubblers.

**Comment by george-basco** [comment]
Looks great!  Now for the truly nitpicky.  Spacing for "Don't Stop" needs adjustment just for the light groups.  The other circuit types all look fine:

![image](https://user-images.githubusercontent.com/70659449/92947960-8b79f800-f40d-11ea-9b7b-2500cd5213ae.png)

**Comment by rstrouse** [comment]
Hmm... do a hard refresh of your browser.  Sometimes the css cache isn't cleared on change.  I'll add a couple of rems to the header width to handle the variance and force the text to a single line.

**Comment by rstrouse** [comment]
Alright I pushed the change so it cannot wrap and gave it a bit more space.

**Comment by george-basco** [comment]
The latest code definitely fixes the issue.  Thanks!

**Comment by theRealMCQwerty** [comment]
> There is one thing to note about lights and "Show In Features." They are odd, while you can create features and circuits that do not appear on the interface, you cannot create lights that do not show on the interface. With lights, the Show In Features checkbox determines whether it shows up in the features list or on the lights list. This is because some of the IntelliBrite based control elements are both lighting and feature. This includes the lighted water feature equipment like MagicStream and the bubblers.

@rstrouse I have a bubbler so I decided to try this. I checked the 'show as feature' box on the circuit that controls the bubbler light transformer. It now shows a single toggle button in the features AND the bubbler light control shows in the 'lights' section.
I suppose this is as I expected but I don't see the point. I get your comment that a bubbler is a feature as well as a light but the 'feature' part of it is really the valve that controls the flow of water to it. What is the point in having a on/off switch on the features section when the full light controls are only in the 'lights' section? Additionally, I can do this for _any_ light circuit (designate it as a feature that is) not just one of the magicstream type items. 
Just trying to unpack your comment a bit and understand what you meant....

**Comment by george-basco** [comment]
This is a dashPanel thing (looks like njspc has all the right data), but related to multiple bodies, so posting here.  Under the heaters config, the body name is not displayed correctly.  In this case "Body 1" should be **Pool** and "Body 2" should be **Spa**.

![image](https://user-images.githubusercontent.com/70659449/93005300-b2672580-f504-11ea-8824-847867692064.png)

**Comment by rstrouse** [comment]
Yeah this is showing the un-mapped value.  We weren't sure which order it all came in with a dual body system.  Out of curiosity do you have a manual heat setting on both bodies or does it exist only on the second body?  You will see this in the OCP bodies page.

**Comment by george-basco** [comment]
They both have manual heat setting:

![image](https://user-images.githubusercontent.com/70659449/93024149-0cb9c200-f5a9-11ea-9bf5-ebff2273038e.png)

**Comment by rstrouse** [comment]
Nope that is not what I mean.  Manual heat is a goofy name for when I turn on that body manually. the heater automatically gets set to the gas heater.  You will find this setting on the bodies configuration.
![image](https://user-images.githubusercontent.com/47839015/93024268-01b36180-f5aa-11ea-97fc-5bad8aa97345.png)

**Comment by george-basco** [comment]
Haven't configured heaters yet, so trying to decipher this stuff.

On the web interface I see exactly what you show above.  Spa Manual Heat option only shows for the Spa body.  However, if I try to set the option, it doesn't stick.  Maybe this is because something isn't configured elsewhere yet?

On the IOS app and OCP, there is no such option on the Bodies configuration pages.  Given that it doesn't seem to do anything on the web interface, makes sense that it doesn't show on OCP or IOS app.

**Comment by rstrouse** [comment]
In the options section there is also a "Manual Heat" setting.  Perhaps it has something to do with that.  The options is bizarre at best but it will tell us whether it actually treats the dual body system as a "Spa" or not.   Right now it looks like you can configure either body to be the pool or you could configure it to be 2 pools.

**Comment by george-basco** [comment]
I turned on the "Manual Heat" option under "General Settings -> Advanced Settings".  This didn't change anything.  I still don't see a "Spa Manual Heat" option under the body configurations page in either the OCP or IOS app interface.  The option does show up in the web interface, but doesn't seem to do anything as I noted before.

Via the "installation setup" on the OCP, I can set the body type.  In this case I made both of them type "spa"
![image](https://user-images.githubusercontent.com/70659449/93025917-edc22c80-f5b6-11ea-864a-5c7e3fba4a98.png)

Or both type "pool"
![image](https://user-images.githubusercontent.com/70659449/93025986-632dfd00-f5b7-11ea-9ec7-35d090efe895.png)

**Comment by rstrouse** [comment]
Omg, there is a body type setting.  Shoot a screenshot of that.  It appears that the manual heat setting is dependent on this body type and that is the determination for the pool or spa or pool, pool or spa spa aliases for the body attachment.

**Comment by george-basco** [comment]
Is this what you want?

![IMG_0602](https://user-images.githubusercontent.com/70659449/93029414-81075c00-f5cf-11ea-87ad-26b17877dc07.jpg)
![IMG_0599](https://user-images.githubusercontent.com/70659449/93029416-8369b600-f5cf-11ea-9e35-97e105fd0e80.jpg)
![IMG_0600](https://user-images.githubusercontent.com/70659449/93029420-85337980-f5cf-11ea-8174-57a59160305a.jpg)
![IMG_0601](https://user-images.githubusercontent.com/70659449/93029422-86fd3d00-f5cf-11ea-85a7-0e868ac2b0e9.jpg)

**Comment by rstrouse** [comment]
Well look at that.  I had no idea.  I am scraping those bytes and didn't do anything with them.  I had just assumed that the body type would be pool for both bodies in an i10D.

**Comment by rstrouse** [comment]
Alright I added support for body types.  Turns out I can do this with an i10PS as well.  We almost had this correct.  Do a pull then reload your config so it picks up the type descriptions.

**Comment by george-basco** [comment]
Sorry if I'm missing it, but don't see your new changes.

**Comment by rstrouse** [comment]
Sorry I didn't hit the push.  It was waiting for me.  It's up now.

**Comment by george-basco** [comment]
Ok, bodies are getting displayed correctly now.  When playing around with the heater settings I noticed an odd behavior.  If I change the "Body" setting for a heater in dashPanel, the change does get made (OCP sees it), but dashPanel reverts back to the previous setting.

Here's an example.  The starting point is to have two heaters, "gas pool" has body = pool, and "gas spa" has body = spa:
![image](https://user-images.githubusercontent.com/70659449/93121153-4659fe00-f679-11ea-9389-4599e5487a5f.png)
![image](https://user-images.githubusercontent.com/70659449/93121213-540f8380-f679-11ea-8c4c-f2f89afa3635.png)

This is the starting point for the replay capture.

After the reload completes, I change "gas pool" body to spa, and "gas spa" body to pool.   When I click the "save heater" button to save the config, the body reverts back to previous state.  This happens for both saves.  However, OCP and web interface both register the changes as expected:
![image](https://user-images.githubusercontent.com/70659449/93121492-c54f3680-f679-11ea-81e5-5319de2bb4b4.png)
![image](https://user-images.githubusercontent.com/70659449/93121527-d1d38f00-f679-11ea-8d99-e00edbe441ab.png)

So making the change through dashPanel is working correctly, but dashPanel state doesn't reflect it.  If I "reload config", the state will show correctly to match what OCP and web interface are showing.

One additional oddity is that both OCP and web interface don't allow the body to be changed.  They show the second body but greyed out.  The IOS app does allow the change to happen like dashPanel.

Here's what the web interface looks like, showing the greyed out pool option:
![image](https://user-images.githubusercontent.com/70659449/93123170-8c649100-f67c-11ea-89d6-d22b6fa588db.png)

Not an important action to begin with and given all the strangeness in behaviors, can't imagine any of this matters.  But just wanted to point it out.

Here is the replay that consists of the reload and the two body changes: 
[replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/5220212/replay.12.zip)

**Comment by rstrouse** [comment]
I'll have a look at the replay.  I don't experience the same results so it could be the 168 messages coming out of series and we are picking them up as gospel.  It is odd that sometimes the web app just broadcasts its current settings after another controller changes them.  I have seen the wireless remote then have to ask for the whole series over again.  This would that our response includes the goofy settings from the web app and the final results haven't been tabulate it.

Maybe it needs a bit of that or it could be sumpin simpler like we only store it if it is a single or shared system.  I'll have a look.

**Comment by rstrouse** [comment]
Well that was none of the above.  Turns out it was reliant on the controllers on the bus to ask for the body setting.  In my case the wireless controller seems to always chime in.  Anyhow it now persists the value without the assistance of any other controllers on the bus.

**Comment by tagyoureit** [comment]
Will this cause a problem with findByObject if a users sends a PUT /state/body/heatMode (or setPoint, etc) with `{name: 'pool', heatMode: 'heater'}` and both body types are set as pool?  

Maybe just an edge case and the answer is "don't do that" but was thinking about it as I was reading this thread.

**Comment by rstrouse** [comment]
No the underlying code is still mapped to pool/spa for shared systems.  The desc field is what changes.  The one weirdo thing is that pool and spa are misnomers they really are body 1 and body 2.  They can be flipped around in IntelliCenter and you can have two pools or two spas on a PS or D model.  Is it an overall problem?  Probably not since the identifiers are really pointing at body 1-4.  For those who are calling "pool" or "spa" (not me either) they will always get body1 when sending pool or body2 when sending "spa."  Not sure if you can flip these or set it up that way in IntelliTouch.

The identifiers are important though cuz when mapping a heater to the body is what determines which J connector is tripped to turn on the specified heater.

**Comment by rstrouse** [comment]
Oh one more thing.  On shared systems there is a fourth options that maps to body1 or body2.  If you have them mapped to Pool/Pool or Spa/Spa that is how it appears.  If you have it set up like the default it is Pool/Spa.  These are equipment mappings which show which body or bodies the equipment is available on.  This is true for chlorinators, chem controllers, and intake/return valves as well.

**Comment by george-basco** [comment]
Options for a Heat Pump don't exist in the other interfaces.  Adding a heat pump via OCP or web interface doesn't offer any additional options.  When viewed in dashPanel, the options below are shown but cannot be set without error.

![image](https://user-images.githubusercontent.com/70659449/93145625-ac5a7b80-f6a1-11ea-9f79-93d1a3285819.png)
![image](https://user-images.githubusercontent.com/70659449/93145606-9b116f00-f6a1-11ea-94b5-7876f18f50ee.png)

**Comment by rstrouse** [comment]
Actually, these settings did exist for a Heat Pump on the web app at one time.  However, they removed the heat pump from the selection recently and added hybrid.  What is really odd about this is that there is a different list on the Mobile App.  Also if you look at the connections these are addressed on the RS485 bus.  What is really odd is that the mobile app uses a different type id than the web app.  Do you have a heat pump wired into your system?  I am looking for someone who has one of these to ferret it all out.

**Comment by george-basco** [comment]
I don't currently have a heat pump, although I'm pretty sure I'll be adding one at some point.

FYI, I see the same set of heater types in dashPanel, interllicenter.com web interface, and IOS app:
Gas, Solar, UltraTemp, MasterTemp, MAX-E Therm, Hybrid, Heat Pump.
All three interfaces show the same 7 types.

However, OCP and ICP (running 1.040) are only showing Gas, Solar, Heat Pump, UltraTemp, Hybrid.

I did see all 7 on ICP when it was running 1.047.

Anyway, given that Pentair can't be consistent and that these are one time setup issues, obviously not a big deal.

**Comment by rstrouse** [comment]
Yeah it will b good 2 have the actual equipment.  I also have a feeling that the heaters will need a revisit if they ever release 1.045 or 1.047

**Comment by george-basco** [comment]
Water Sensor 2 is not being read.  Looks like there is some dual equipment logic needed around here:

https://github.com/tagyoureit/nodejs-poolController/blob/f0ef553edeb089e8acaff3b8f80906d9ee0b3376/controller/comms/messages/status/EquipmentStateMessage.ts#L399

In my case byte 15 does have the Water Sensor 2 data.   Haven't looked at the solar sensors..

**Comment by rstrouse** [comment]
Pull Next, this should be fixed.  It was previously only looking for expansion panels to populate this value. It now checks to see if we have a shared or dual system as well.  waterSensor2 is not valid in shared or single body systems without an expansion.

**Comment by george-basco** [comment]
With 1.047 firmware released, I looked at differences between the various interfaces.  All the interfaces show the same 7 types of heaters.  Comparing OCP running 1.047, dashPanel (latest -next), Web Client Version: 1.048 b8 Jun 9, 2020, IOS app 2.0.133

Gas: OCP and IOS app have "Cooldown Delay" option; dashPanel and web do not; 
Solar: same for all 4 interfaces
Heat Pump: dashPanel has options for "Address", "Differential Temp", and "Nocturnal Cooling.  OCP, IOS app, and web have none of these
UltraTemp: OCP has option for "Cooling Enabled";  dashPanel shows "Differential Temp" and "Nocturnal Cooling"; IOS app and web have "Differential Cooling" and "Cooling Enabled"
MasterTemp: same
Max E-Therm: same
Hybrid: same except IOS app adds a "Heat Mode" option

**Comment by george-basco** [comment]
I noticed sys.general.options.pumpDelay is not getting read correctly on a reload.
https://github.com/tagyoureit/nodejs-poolController/blob/41ab9efb0eb275bd97edf59d56d315f86c5438ba/controller/comms/messages/config/OptionsMessage.ts#L46

instead of extracting byte 37, it should be 30.  This seems to fix it on my system and matches what's here:
https://github.com/tagyoureit/nodejs-poolController/blob/41ab9efb0eb275bd97edf59d56d315f86c5438ba/controller/boards/IntelliCenterBoard.ts#L931-L934


It also looks like this code is not correct:
https://github.com/tagyoureit/nodejs-poolController/blob/41ab9efb0eb275bd97edf59d56d315f86c5438ba/controller/boards/IntelliCenterBoard.ts#L963-L966
I think the numbers 36,39 should be 37,40?

This time to match what's here:
https://github.com/tagyoureit/nodejs-poolController/blob/41ab9efb0eb275bd97edf59d56d315f86c5438ba/controller/comms/messages/config/OptionsMessage.ts#L48

The manual heat option is not exposed in dashPanel for me, so couldn't check this.

**Comment by george-basco** [comment]
> Haven't configured heaters yet, so trying to decipher this stuff.
> 
> On the web interface I see exactly what you show above. Spa Manual Heat option only shows for the Spa body. However, if I try to set the option, it doesn't stick. Maybe this is because something isn't configured elsewhere yet?
> 
> On the IOS app and OCP, there is no such option on the Bodies configuration pages. Given that it doesn't seem to do anything on the web interface, makes sense that it doesn't show on OCP or IOS app.

Btw, I think I'm a bit less confused about the Spa Manual Heat option.  Don't know if it's because of my dual equipment config, but this option no longer seems to be a per-body option and has moved to a global option under "General Settings -> Advanced Settings" (on web interface).  dashPanel should add this to the general settings tab.  Might need to change the name of "delays" or split it out with the manual op priority option.

Currently dashPanel shows "spa manual heat" option for spa type bodies under bodies tab.  The setting does seem to stick, once you set it, but it has no effect.   OCP and IOS app do not show a per-body option.  The web interface does still show it, but the option doesn't work, so guessing a bug that it shows up at all.

**Comment by george-basco** [comment]
The i10D, like the i10PS, should have 11 circuits.  9 AUX + pool + spa.  Current definition has 10.

https://github.com/tagyoureit/nodejs-poolController/blob/41ab9efb0eb275bd97edf59d56d315f86c5438ba/controller/boards/IntelliCenterBoard.ts#L111

**Comment by rstrouse** [comment]
@george-basco,

I posted changes to support the 1.047 firmware.  You will need both njspc and -dashPanel.

**Comment by george-basco** [comment]
Part of https://github.com/tagyoureit/nodejs-poolController/issues/207#issuecomment-694510153 is still a problem:

> 
> It also looks like this code is not correct:
> https://github.com/tagyoureit/nodejs-poolController/blob/41ab9efb0eb275bd97edf59d56d315f86c5438ba/controller/boards/IntelliCenterBoard.ts#L963-L966
> 
> 
> I think the numbers 36,39 should be 37,40?
> This time to match what's here:
> https://github.com/tagyoureit/nodejs-poolController/blob/41ab9efb0eb275bd97edf59d56d315f86c5438ba/controller/comms/messages/config/OptionsMessage.ts#L48
> 
> I also was able to verify this change does seem to work correctly on my system..

**Comment by rstrouse** [comment]
Funny, I documented it that way originally then probably looked at the 30 message documentation to set the 168 command.

**Comment by rstrouse** [comment]
I am closing this open a new issue if there are different problems.
